### ISSUE TITLE
Some unicode takes us down the rabbit hole: कॉम

### ISSUE THREAD

**Issue by invisig0th** [vertexproject/synapse#170]
The unicode कॉम  is the unicode representation of an IDNA encoded IANA TLD and breaks synapse because linux/python doesn't see all 3 chars as "word chars" even though they are....

must we do *everything* ourselves? sigh...

**Comment by vEpiphyte** [comment]
@invisig0th is this still an issue with IDNA / TLD changes that have happened since this issue was created?

**Comment by vertexmc** [comment]
I still get `BadTypeValu`

**Comment by MichaelSquires** [comment]
The referenced unicode (कॉम) is actually three unicode characters (See [this page](http://www.utf8-chartable.de/unicode-utf8-table.pl?start=2304&number=128) for code point reference):
1. `b'\xe0\xa4\x95'` (U+0915: DEVANAGARI LETTER KA)
1. `b'\xe0\xa5\x89'` (U+0949: DEVANAGARI VOWEL SIGN CANDRA O)
1. `b'\xe0\xa4\xae'` (U+092e: DEVANAGARI LETTER MA)

I checked each individual character against a simple regex of `\w`. The results are that U+0915 and U+092e both match while U+0949 doesn't match. Results below:

```
>>> re.match('(?u)\w', b'\xe0\xa4\x95'.decode('utf8'))
<_sre.SRE_Match object; span=(0, 1), match='क'>
>>> re.match('(?u)\w', b'\xe0\xa4\xae'.decode('utf8'))
<_sre.SRE_Match object; span=(0, 1), match='म'>
>>> re.match('(?u)\w', b'\xe0\xa5\x89'.decode('utf8'))
>>>
```

Given the description of U+0949, I believe that the reason the whole unicode string doesn't match is because U+0949 really isn't a word character but some sort of modifier. I checked other characters that were also described as "VOWEL SIGN" and none of those matched either.

In a nutshell, I think this language (Hindi) just happens to have characters that regex doesn't recognize as word characters and it's causing the [`fqdnre`](https://github.com/vertexproject/synapse/blob/master/synapse/models/inet.py#L67) to break.

Thoughts?

**Comment by MichaelSquires** [comment]
And, right after that *amazing* comment I just made (:smile:), I found a stack overflow question from 5 years ago about the EXACT SAME problem. (To be fair, I wouldn't have found it if I didn't know the problem was with vowel signs.) The answer is that python's `re` module *IS* broken but the desired functionality works in `regex`. Proof is in the pudding below.

```
>>> fqdnre = regex.compile(r'^[\w._-]+$', regex.U)
>>> fqdnre.match(b'\xe0\xa4\x95\xe0\xa5\x89\xe0\xa4\xae'.decode('utf8'))
<regex.Match object; span=(0, 3), match='\u0915\u0949\u092e'>
>>>
```

Any objections to moving to `regex` instead of python's builtin `re` module? It seems to be API compatible.
